### ISSUE TITLE
Reconfiguring a model after training

### ISSUE THREAD

**Issue by jfsantos** [fchollet/keras#79]
Certain layers, such as Dropout and BatchNormalization, are supposed to be used only during training. It would be nice if we could disable them during testing or for using a model in production. This could be done by changing the connections between layers or replacing a layer by an `Identity` layer. Any ideas?

**Comment by jfsantos** [comment]
Would something like this work?

```python
test_layers = []
for layer in model.layers:
    if type(layer) not in [BatchNormalization, Dropout]:
        test_layers.append(layer)
        if len(test_layers) > 1:
            layer.connect(test_layers[-2])
model.layers = test_layers
model.compile(...)
```

If that's the case, we could only add a utility function similar to this to remove layers of arbitrary types (provided by the user in a list) and redo the connections.

**Comment by fchollet** [comment]
You can't replace Dropout with an identity layer, because Dropout does not behave like an identity layer at test time. It recalibrates the input by a factor  `retain_proba`, which is necessary since the weights were learned with only a fraction  `retain_proba` of the input activated at a time. 

In general calling a layer with `train=False` will always make it production-compliant. This is what happens when you use the methods `test`, `evaluate`,  `predict_proba` or `predict_classes` of a model.

**Comment by jfsantos** [comment]
Right, I overlooked this detail. Thanks for the clarification! I didn't want to have batch normalization during test, but it's just a matter of using `batch_size=1` in `predict_proba` and friends.

**Comment by fchollet** [comment]
Statistically, if the weights are learned with batch normalization, wouldn't it be necessary to keep batch normalization at test time, with the same batch size? Otherwise, wouldn't it affect the distribution of the inputs and therefore impact performance?

**Comment by jfsantos** [comment]
According to the [paper](http://arxiv.org/pdf/1502.03167v3.pdf) (page 4, section 3.1), during inference you should use the population statistics instead of the mini-batch statistics.

**Comment by fchollet** [comment]
Ok, interesting. In that case maybe we should update the `train=False` behavior of the layer to reflect this. It would simply return its input unchanged, right?

**Comment by jfsantos** [comment]
My prior understanding of this was wrong. Actually, the layer works the same way, but the normalization part is done using the mean and variance of the training data. The variance is computed using the unbiased variance estimate over size m mini-batches in the training data: Var[x] = m/(m-1) * E_B(\sigma^2_B). So, instead of being stochastic (depending on the batch content), the batch normalization layer is actually a linear transform over the activations during inference.

**Comment by fchollet** [comment]
Ok, I see. So we should indeed update the test behavior (train=False) of the layer to use the mean and variance of the training data (learned at train=True).

**Comment by jfsantos** [comment]
Maybe obvious, but this also means we have to save these as model attributes (so we can serialize them with the layers as well).

**Comment by isaacgerg** [comment]
@jfsantos Has this been fixed?

**Comment by jfsantos** [comment]
@isaacgerg this has been fixed a long time ago and I forgot to mark the issue as solved. Sorry!

**Comment by Nimi42** [comment]
At first I want to say I'm sorry that I post this here, but I could not find a solution to my problem on the net.

Let's say I want to train a GRU and because I need stateful=true the batch-size has to be known beforehand and the timesteps also have to be known in advance

Using the functional API I would have an Input as follows:

```python
input_1 = Input(batch_shape=(batch_size, time_steps, features))
```

But when I evaluate the model I don't want to pass my test data in batches.  My
solution I time is to load the save model and rebuild it with 

```python
input_1 = Input(shape=(None, num_input_dim))
```

The must be a better way to redefine the input shape. Can somebody help me out here
please.
